lib/ops/upsample.ts

```diff
@@ -3,33 +3,104 @@
+    (a: number, b: boolean): number;
+  }
+  interface GetOriginalCoordinateFunc {
+    (xResized: number, xScale: number, lengthResized: number, lengthOriginal: number, roiStart: number,
+     roiEnd: number): number;
+  }
+}
+
 export abstract class Upsample implements Operator {
+  constructor(protected opset: number) {}
+
   abstract run(inferenceHandler: InferenceHandler, inputs: Tensor[]): Tensor[]|Promise<Tensor[]>;
 
-  initialize(attributes: Attribute): void {
-    this.mode = attributes.getString('mode', 'nearest');
-    this.scales = attributes.getFloats('scales');
+  initialize(attributes: Attribute, node: Graph.Node, graph: Graph): void {
+    this.isResize = (this.opset >= 10);
 
-    if (this.mode !== 'nearest' && this.mode !== 'linear') {
+    this.mode = attributes.getString('mode', 'nearest');
+    if (this.mode !== 'nearest' && this.mode !== 'linear' && (this.opset < 11 || this.mode !== 'cubic')) {
       throw new Error(`unrecognized mode: ${this.mode}`);
     }
 
-    if (this.mode === 'linear' && this.scales.length !== 2 && this.scales.length !== 4) {
-      throw new Error(`only support 2-D or 4-D upsampling for linear mode`);
+    if (this.opset < 9) {
+      this.scales = attributes.getFloats('scales');
+      scalesValidataion(this.scales, this.mode, this.isResize);
+    }
+
+    this.extrapolationValue = attributes.getFloat('extrapolation_value', 0.0);
+
+    this.coordinateTransformMode =
+        this.opset > 10 ? attributes.getString('coordinate_transformation_mode', 'half_pixel') : 'asymmetric';
+    if ([
+          'asymmetric', 'pytorch_half_pixel', 'tf_half_pixel_for_nn', 'align_corners', 'tf_crop_and_resize',
+          'half_pixel'
+        ].indexOf(this.coordinateTransformMode) === -1) {
+      throw new Error(`coordinate_transform_mode '${this.coordinateTransformMode}' is not supported`);
+    }
+    this.useExtrapolation = this.needRoiInput = (this.coordinateTransformMode === 'tf_crop_and_resize');
+
+    this.nearestMode =
+        (this.mode === 'nearest' && this.opset >= 11) ? attributes.getString('nearest_mode', 'round_prefer_floor') : '';
+    if (['round_prefer_floor', 'round_prefer_ceil', 'floor', 'ceil', ''].indexOf(this.nearestMode) === -1) {
+      throw new Error(`nearest_mode '${this.nearestMode}' is not supported`);
+    }
+
+    this.cubicCoefficientA = attributes.getFloat('cubic_coeff_a', -0.75);
+    this.excludeOutside = attributes.getInt('exclude_outside', 0) !== 0;
+    if (this.excludeOutside && this.mode !== 'cubic') {
+      throw new Error('exclude_outside can be set to 1 only when mode is CUBIC.');
+    }
+
+    this.useNearest2xOptimization = (this.opset < 11) ?
+        true :
+        (this.mode === 'nearest' && this.coordinateTransformMode === 'asymmetric' && this.nearestMode === 'floor');
+
+    if (this.opset > 10) {
+      this.roiInputIdx = 1;
+      this.scalesInputIdx = 2;
+      this.sizesInputIdx = 3;
+    } else if (this.opset === 9) {
```

Comment:
I ran into an issue when trying to use this when this.opset was 10. In my particular use case, adding `|| this.opset === 10` caused the inference to run as expected, but I am not an expert regarding this codebase.